apiserver/facades/client/action/operation.go

```diff
@@ -5,94 +5,169 @@ package action
 func (a *ActionAPI) EnqueueOperation(ctx context.Context, arg params.Actions) (params.EnqueuedActions, error) {
-	operationId, actionResults, err := a.enqueue(ctx, arg)
-	if err != nil {
-		return params.EnqueuedActions{}, err
-	}
-	results := params.EnqueuedActions{
-		OperationTag: names.NewOperationTag(operationId).String(),
-		Actions:      actionResults.Results,
+	if err := a.checkCanWrite(ctx); err != nil {
+		return params.EnqueuedActions{}, errors.Capture(err)
 	}
-	return results, nil
-}
 
-func (a *ActionAPI) enqueue(ctx context.Context, arg params.Actions) (string, params.ActionResults, error) {
-	if err := a.checkCanWrite(ctx); err != nil {
-		return "", params.ActionResults{}, errors.Trace(err)
+	if len(arg.Actions) == 0 {
+		return params.EnqueuedActions{}, apiservererrors.ServerError(errors.New("no actions specified"))
 	}
 
-	return "", params.ActionResults{}, errors.NotSupportedf("actions in Dqlite")
-}
+	const leader = "/leader"
+	actionResults := make([]params.ActionResult, len(arg.Actions))
+	actionResultByUnitName := make(map[string]*params.ActionResult)
+	runParams := make([]operation.RunArgs, 0, len(arg.Actions))
+	for i, action := range arg.Actions {
+		taskParams := operation.TaskArgs{
+			ActionName:     action.Name,
+			Parameters:     action.Parameters,
+			IsParallel:     action.Parallel == nil || *action.Parallel, // default is parallel
+			ExecutionGroup: zeroNilPtr(action.ExecutionGroup),
+		}
+
+		if strings.HasSuffix(action.Receiver, leader) {
+			receiver := strings.TrimSuffix(action.Receiver, leader)
+			actionResultByUnitName[action.Receiver] = &actionResults[i]
+			runParams = append(runParams, operation.RunArgs{
+				Target:   operation.Target{LeaderUnit: []string{receiver}},
+				TaskArgs: taskParams,
+			})
+			continue
+		}
+
+		unitTag, err := names.ParseUnitTag(action.Receiver)
+		if err != nil {
+			actionResults[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+		actionResultByUnitName[unitTag.Id()] = &actionResults[i]
+		runParams = append(runParams, operation.RunArgs{
+			Target: operation.Target{
+				Units: []unit.Name{unit.Name(unitTag.Id())},
+			},
+			TaskArgs: taskParams,
+		})
 
-// ListOperations fetches the called actions for specified apps/units.
-func (a *ActionAPI) ListOperations(ctx context.Context, arg params.OperationQueryArgs) (params.OperationResults, error) {
-	if err := a.checkCanRead(ctx); err != nil {
-		return params.OperationResults{}, errors.Trace(err)
 	}
 
-	var receiverTags []names.Tag
-	for _, name := range arg.Units {
-		receiverTags = append(receiverTags, names.NewUnitTag(name))
+	// If no valid run params (all receivers invalid), do not call service; return per-action errors.
+	if len(runParams) == 0 {
+		return params.EnqueuedActions{Actions: actionResults}, nil
 	}
-	for _, id := range arg.Machines {
-		receiverTags = append(receiverTags, names.NewMachineTag(id))
+
+	result, err := a.operationService.Run(ctx, runParams)
```

Comment:
question: what is the purpose of having the `Run` service method handle all of `params.Actions` at one time? There isn't a transactional reason to do so I can see. Each element of `params.Actions.Actions` is unique.